### PR TITLE
[aws-lambda] Narrow AuthResponseContext type

### LINKED ISSUE #42418: AuthResponseContext in aws-lambda should not accept props of type any
The context properties of an AuthResponseContext (types/aws-lambda/common/api-gateway.d.ts) can be of types string, bools and numbers hence **any** is misleading.

See: https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-lambda-authorizer-output.html

The interface AuthResponseContext should not accept properties of type **any**, instead it should accept properties of type **string | boolean | number**.

### PR DESCRIPTION
**Pull request by furudean**
Fixes #42418

> You can access the stringKey, numberKey, or booleanKey value (for example, "value", "1", or "true") of the context map in a mapping template by calling $context.authorizer.stringKey, $context.authorizer.numberKey, or $context.authorizer.booleanKey, respectively. The returned values are all stringified. Notice that you cannot set a JSON object or array as a valid value of any key in the context map.

From https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-lambda-authorizer-output.html

### CODE REVIEW

**Comment by simonbuchan** [comment]
Docs are a bit dumb here, there's been an attempt to narrow before is #33437 that was reverted since the cognito authorizer will give you an object in `event.authorizer.claims`. This can be fixed by splitting the authorizer (output) and proxy handler (input) types, but that's a bigger breaking change.

I've put some thoughts about where I think this should go in the related issue #34059, any feedback?

**Comment by c-bandy** [comment]
Ah! Shame it isn't as straight-forward as documented.

What do you think about using a union type to represent a generic `context`? This could cover the custom authorizer context and the cognito context property.

**Comment by c-bandy** [comment]
On #34059, I basically agree. I think your last solution is as elegant as it gets. 

If not for that, at the very least we should differentiate the `APIGatewayProxyEvent.requestContext`, `CustomAuthorizerResult.context` and `CustomAuthorizerEvent.requestContext` types.

**Comment by simonbuchan** [comment]
Yeah, I'm happy narrowing the authorizer result context, it should only catch actual errors (unless someone's playing tricks with `toString()` / `toJSON()`? They knew what they were getting into.)

I also have #42420 open now which will conflict with this if it goes in. If I'm bringing in a new type anyway I can make these changes at the same time?

**Comment by c-bandy** [comment]
I'll wait for your PR to be merged until I make any changes here

**Comment by c-bandy** [comment]
@simonbuchan sorry, looks like i completely parsed your question wrong. feel free to do so!

**Review by StefH — APPROVED** [review]

**Comment by darbio** [comment]
Isn't this going to be addressed by @simonbuchan in #42420? If so this PR should **not** be merged.

**Review by simonbuchan — CHANGES_REQUESTED** [review]
More relevantly, a very similar PR #33437 has previously had to have been rolled back, so whether this gets added to my PR or not (I was just commenting that it would conflict) this shouldn't be merged.
I'm going to mark this request changes so it doesn't slip through.

**Comment by simonbuchan** [comment]
... @c-bandy you thumbed that up so fast it was there before the page loaded after my post 😲

**Comment by c-bandy** [comment]
@simonbuchan I was already here writing essentially the same message you were, but you were faster than me :P

**Comment by simonbuchan** [comment]
Since it's been open for a week and at least nobody seems to be complaining about it, I'll throw the changes in.

**Comment by simonbuchan** [comment]
Let me know if the latest changes in #42420 look like they fix this for you.

**Comment by c-bandy** [comment]
I havent had time to test your changes but the source code looks good to me

**Comment by c-bandy** [comment]
Hey @simonbuchan, I tried testing this in one of our repos, it looks like the issue raised here is not solved by your work so far, I can still put `any` inside the context for a custom authorizer. Am I expected to use a different type here?

image

**Comment by simonbuchan** [comment]
Ah, yes, by default my PR doesn't narrow the response context type, it probably should. You can instead declare a type that describes the context exactly, then use that with `APIGatewayAuthorizerWithContextResult<MyContext>`, and matching `APIGatewayProxyWithContextEvent<MyContext>`.

I can quickly add narrowing the response type in my PR on my Monday, which should be before a maintainer merges it, since I'm UTC+12, but I might forget. Won't be a huge problem to PR just that though.

**Pull request closed without merging**
